cou/steps/analyze.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2023 Canonical Limited.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Functions for analyzing an OpenStack cloud before an upgrade."""
+from __future__ import annotations
+
+import logging
+from collections import defaultdict
+from dataclasses import dataclass, field
+from io import StringIO
+from typing import Any, Iterable
+
+from juju.client._definitions import ApplicationStatus
+from ruamel.yaml import YAML
+
+from cou.utils.juju_utils import (
+    async_get_application_config,
+    async_get_full_juju_status,
+    extract_charm_name_from_url,
+)
+from cou.utils.openstack import CHARM_TYPES, get_os_code_info
+
+
+@dataclass
+class Analysis:
+    """Analyze result."""
+
+    apps: Iterable[Application]
+
+    @classmethod
+    async def create(cls) -> Analysis:
+        """Analyze the deployment before planning."""
+        logging.info("Analyzing the OpenStack deployment...")
+        apps = await Analysis._populate()
+
+        return Analysis(apps=apps)
+
+    @classmethod
+    async def _populate(cls) -> set[Application]:
+        """Generate the applications model."""
+        juju_status = await async_get_full_juju_status()
+        model_name = juju_status.model.name
+        apps = {
+            Application(
+                name=app,
+                status=app_status,
+                config=await async_get_application_config(app),
+                model_name=model_name,
+            )
+            for app, app_status in juju_status.applications.items()
+        }
+        return apps
+
+    def __str__(self) -> str:
+        """Dump as string."""
+        return "\n".join([str(app) for app in self.apps])
```

Comment:
not-blocking: `os.linesep` instead of "\n"

Reply:
ok